Support GSM 07.10 muxing protocol for generic GSM modem

jukkar opened the pull request:
This one supports GSM muxing. Tested with SIM808 modem.

mike-scott commented:
@jukkar Yes, I had a thought to perhaps split the modem drivers out into separate parts:

- Modem initialization applicable to any network interface (PPP, AT-command socket-based offloading, AT-command netcontext based offloading).
- The AT-command socket-based code implementing those offload APIs
- The AT-command netcontext code implementing those offload APIs

Still deciding on a path here.

weinholtendian commented:
> My implementation is not using virtual UARTs as Zephyr does not have that kind of concept and I did not want to start to implement this kind of big feature just for muxing.

@jukkar There are several UARTs already that aren't tied to a physical UART chip, e.g. USB CDC and RTT UART. What distinguishes virtual UARTs from other UARTs?

jukkar commented:
> What distinguishes virtual UARTs from other UARTs?

Probably nothing, I just need to investigate this route more.

GeorgeGkinis commented:
Although I am not adequately technically adept to implement the 07.10 protocol, I suspect that implementing a virtual UART driver could be the least intruding feature.
 
- When initialising a vUART driver a physical UART would be selected. 
- It would expose a multitude of vUARTs with the same API as a hardware one.
- Then everything that uses UART could write-read from it.
     - Selecting a vUART for **PPP**
     - Using **non standard AT commands** or a custom library per module instead of a rigid API
     - Selecting a NMEA library to parse incomming data (assign it to a vUART)
     - vUART _could_ be reused to open many vUART channels to other MCU's 
     - leverage GSM 07.10 for other purposes, as it has been done in the case of [bluetooth](https://docs.zephyrproject.org/1.13.0/subsystems/bluetooth/rfcomm-pics.html).

Of course I am already delighted that at one point I can use 07.10 for my SIMCOM modules in any way, just my 2 cents.

I am curious about the issue of the **SIM808** not answering after channel initialization.
Can I just clone and compile an example code to experiment myself?

Maybe an example as **/jukkar/zephyr/tree/master/samples/net/gsm_modem_0710** could be nice so that collaborators stay on the same page.

jukkar commented:
> Maybe an example as **/jukkar/zephyr/tree/master/samples/net/gsm_modem_0710** could be nice so that collaborators stay on the same page.

You can try this PR with existing gsm_modem sample app like this:
```
$ cd samples/net/gsm_modem
$ cmake -B build . -DBOARD=reel_board -DOVERLAY_CONFIG=prj-mux.conf
$ make -C build flash
```
I have connected the modem to reel_board so you would need to adjust the board name etc to have it working in some other board. At least for my SIM808 modem (https://www.adafruit.com/product/2542), one needs to connect four wires (TX, RX, GND and Vio) from modem to the zephyr board.

My `prj-mux.conf` file looks like this:
```
# UART support
CONFIG_SERIAL=y

# GSM modem support
CONFIG_MODEM=y
CONFIG_MODEM_GSM_PPP=y
CONFIG_MODEM_GSM_UART_NAME="UART_1"

# PPP networking support
CONFIG_NET_PPP=y
CONFIG_NET_L2_PPP=y
CONFIG_NET_NATIVE=y
CONFIG_NETWORKING=y

CONFIG_NET_L2_PPP_TIMEOUT=10000

# IPv4 enables PPP IPCP support
CONFIG_NET_IPV4=y
CONFIG_NET_IPV6=n

# Network management events
CONFIG_NET_CONNECTION_MANAGER=y

# Log buffers, modem and PPP
CONFIG_LOG=y
CONFIG_NET_LOG=y
CONFIG_LOG_BUFFER_SIZE=65536
CONFIG_LOG_STRDUP_BUF_COUNT=900
CONFIG_LOG_FUNC_NAME_PREFIX_ERR=y
CONFIG_LOG_FUNC_NAME_PREFIX_INF=y
CONFIG_NET_DEBUG_NET_PKT_ALLOC=y
CONFIG_NET_BUF_POOL_USAGE=y
CONFIG_MODEM_LOG_LEVEL_DBG=y
CONFIG_MODEM_CONTEXT_VERBOSE_DEBUG=y
#CONFIG_NET_PPP_LOG_LEVEL_DBG=y
CONFIG_NET_L2_PPP_LOG_LEVEL_DBG=y
#CONFIG_NET_MGMT_EVENT_LOG_LEVEL_DBG=y
#CONFIG_NET_CONNECTION_MANAGER_LOG_LEVEL_DBG=y
CONFIG_MODEM_GSM_MUX=y

CONFIG_MODEM_GSM_MUX_LOG_LEVEL_DBG=y
#CONFIG_MODEM_GSM_MUX_VERBOSE_DEBUG=y
CONFIG_NET_SHELL=y
CONFIG_MODEM_SHELL=y
CONFIG_MODEM_GSM_SIMCOM=y
#CONFIG_MODEM_GSM_MUX_DELAY_AFTER_ENABLE=500
#CONFIG_MODEM_GSM_T1_TIMEOUT=2000

CONFIG_KERNEL_SHELL=y
CONFIG_THREAD_MONITOR=y
CONFIG_THREAD_NAME=y
CONFIG_THREAD_STACK_INFO=y
CONFIG_INIT_STACKS=y
CONFIG_DEVICE_SHELL=y
```

jukkar commented:
> Probably nothing, I just need to investigate this route more.

Indeed, after looking briefly the usb implementation, we could probably implement the "virtual UART" similar way as how usb is accessing uarts. It just requires changes to modem iface interface and of course the implementation that hides the muxing behind these virtual uarts.

jukkar commented:
I managed to find the issue where the communication was failing after the cmux handshakes. The UIH frame was incorrectly constructed which caused the modem to drop all the messages that were sent after the handshake.

jukkar committed:
net: ppp: Convert PPP driver to use normal UART APIs

This is needed so that we can support GSM 07.10 muxing protocol.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ppp: Create own workqueue for PPP

In order to avoid sleeping in global workqueue, use own
workqueue for PPP traffic.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: console: uart_mux: Use own workqueue

Instead of global workqueue, use own one so that the users of this
driver can more conveniently use global one. This means GSM modem
can work better with the UART muxing driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
modem: gsm: Initialize the modem UART separately

This is needed if muxing is enabled in which case we must
change the UART to muxing UART after the AT+CMUX command
has succeed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
modem: shell: Add info command

Add an information command that currently will only prints
GSM muxing status.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: gsm_mux: Support GSM 07.10 muxing protocol

Add support to GSM 07.10 muxing protocol which is used to
share the same UART for PPP and AT commands among other things.
This allows e.g., the modem to send SMS and have PPP connection
active at the same time.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: console: uart_mux: Initial UART muxing support

Create support for muxed UARTs which are attached to a real
UART and which use GSM 07.10 muxing protocol to create virtual
channels that can be run on top of the real UART.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: ppp: Enable PPP interface when connection is ready

By default PPP interface is not taken up automatically but only
after the PPP connection to modem is ready.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
drivers: modem: gsm: Add support for GSM 07.10 muxing protocol

Instead of using physical UART to connection to the modem, use
the GSM 07.10 muxing protocol and UART mux driver to create
virtual channels to the modem. This will allow simultaneous
PPP, AT and other type connections to the modem.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Just uploaded a prototype for GSM muxing that I managed to test with SIM808 generic GSM modem (not using offloaded driver).

jukkar committed:
CODEOWNERS: Add GSM and UART mux files owner

Add myself as an owner of these GSM 07.10 mux files.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
If anyone wants to try this, one can use the `samples/net/gsm_modem` sample and set these options
```
CONFIG_GSM_MUX=y
CONFIG_UART_MUX=y
```
The GSM modem will use the muxing automatically.

jukkar commented:
Ping to anyone willing/able to review this, the plan is to get this included in 2.3. The merge window closes in a month but I would like to avoid things reviewed in last week as then we will probably won't get this to 2.3.

weinholtendian commented on drivers/net/ppp.c:
```diff
@@ -176,7 +188,11 @@ static void ppp_change_state(struct ppp_driver_context *ctx,
-		uart_pipe_send(ppp->send_buf, off);
+		u8_t *buf = ppp->send_buf;
+
+		do {
+			uart_poll_out(ppp->dev, *buf++);
+		} while (--off);
```
The final call to `ppp_send_flush` in `ppp_send` can pass `off=0`. If that happens then `off` goes negative, the loop keeps transmitting from outside the buffer.

```suggestion
		while (off--) {
			uart_poll_out(ppp->dev, *buf++);
		};
```

jukkar replied:
Thanks, fixed.

weinholtendian commented:
Got PPP over GSM mux working with a U-blox SARA R412M after fixing the infinite loop in `ppp_send_flush`. Haven't done a full review, but I will continue to test it.

pfl commented on drivers/console/uart_mux.c:
```diff
@@ -0,0 +1,821 @@
+			dev_data->attach_cb = cb;
+			dev_data->attach_user_data = user_data;
+
+			dev_data->dlci = gsm_dlci_create(real_uart->mux,
```
gsm_dlci_create() failure is not checked here. Perhaps do the check before filling in all the other fields and return error if not created?

jukkar replied:
Yes, good point, will fix.

jukkar replied:
I added checks for the return code, unfortunately the other fields are needed and must be set before the function is called as the `gsm_dlci_create()` will return immediately and those fields are used when negotiating the control channel, otherwise we have null ptr access.

pfl commented on drivers/net/ppp.c:
```diff
@@ -698,10 +752,97 @@ static struct net_stats_ppp *ppp_get_stats(struct device *dev)
+			LOG_ERR("Rx buffer doesn't have enough space. "
+				"Bytes pending: %d, written: %d",
+				rx, ret);
+			break;
```
If the ring buffer does not have enough space, this seems to mean that those bytes are lost forever, since the next time this function is called new bytes are read from the fifo and the previous ones in the buffer overwritten. Do we loose sync and get stuck in ppp since the next fifo read will be glued to the end of the previous bytes for which the ring buffer had space?

jukkar replied:
We might loose sync temporarily in this case but ppp will not get stuck as next received sync byte 0x7e will start the frame again.

weinholtendian commented:
@jukkar Is there a plan for how to reinitialize everything if the modem loses its state? This comes up in applications that power off the modem to conserve energy. Just taking the interface down and then back up does not work today. (It didn't before this patch either).

jukkar commented:
> Is there a plan for how to reinitialize everything if the modem loses its state? This comes up in applications that power off the modem to conserve energy. Just taking the interface down and then back up does not work today. (It didn't before this patch either).

This kind of thing was not working earlier either. This needs to be addressed of course but in another PR.

jukkar commented:
Updated according to comments.

jukkar committed:
drivers: gsm_mux: Convert to use k_timeout_t

Convert to use k_timeout_t

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Rebased against latest master. If there are no objections from anybody, I will merge this one tomorrow.

jukkar committed:
doc: net: Add information about GSM 07.10 muxing support

Add description of options that needs to be enabled for GSM
muxing.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Forgot to update the documentation about muxing so one extra commit added for that.
